framework/Optimizers/SimulatedAnnealing.py

```diff
@@ -0,0 +1,718 @@
+    # while in violation of constraints ...
+    tries = 500
+    while not passFuncs:
+      modded = True
+      #  try to find new acceptable point
+      denormed = self.denormalizeData(suggested)
+      suggested, _= self._fixFuncConstraintViolations(suggested, previous) # need info, traj
+      denormed = self.denormalizeData(suggested)
+      self.raiseADebug(' ... suggested new opt {}'.format(denormed))
+      passFuncs = self._checkFunctionalConstraints(denormed)
+      tries -= 1
+      if tries == 0:
+        self.raiseAnError(NotImplementedError, 'No acceptable point findable! Now what?')
+    return suggested, modded
+  ###########
+  # Utility Methods #
+  ###########
+  def _temperature(self, fraction):
+    #return max(0.01,min(1,1-fraction))
+    return max(0.01,1-fraction)
+
+  def _coolingSchedule(self, iter, T0):
+    """TODO write the math here"""
+    #defaults
+    if T0 == None:
+      T0 = 1e4
+
+    if not self._coolingMethod:
+      self._coolingMethod = 'exponential'
+
+    if not self._CoolingParameters:
+      self._CoolingParameters['alpha'] = 0.94
+      self._CoolingParameters['beta'] = 0.1
+      self._CoolingParameters['c'] = 1.0
+      self._CoolingParameters['d'] = 1.0
+
+    type = self._coolingMethod
+    if type == 'linear':
+      beta = self._CoolingParameters['beta']
+      return T0 - iter * beta
+    elif type in ['exponential','geometric']:
+      alpha = self._CoolingParameters['alpha']
+      return alpha ** iter * T0
+    elif type == 'boltzmann':
+      d = self._CoolingParameters['d']
+      return T0/(np.log10(iter + d))
+    elif type == 'fast':
+      c = self._CoolingParameters['c']
+      return np.exp(-c*iter) * T0
+    elif type == 'veryfast':
+      c = self._CoolingParameters['c']
+      return np.exp(-c*iter**(1/len(self.toBeSampled.keys()))) * T0
+    elif type == 'cauchy':
+      d = self._CoolingParameters['d']
+      return T0/(iter + d)
+    else:
+      raise NotImplementedError('cooling schedule type not implemented.')
+
+  def _nextNeighbour(self, rlz,fraction=1,alpha = 0.94,cont = True):
+    """ Perturb x to find the next random neighbour
+        for linear and exponential cooling:
+        .. math::
+
+            fraction = \\frac{iter}{Limit}
+
+            amp = 1-fraction
+
+            delta = \\frac{-amp}{2} + amp * r
+        where :math: `r \sim \mathcal{U}(0,1)`
+
+        for boltzmann cooling:
+        .. math::
+
+            amp = min(\\sqrt(T), \\frac{1}{3*alpha}
+
+            delta = r * alpha * amp
+
+        where :math: `r \\sim \\mathcal{N}(0,1)`
+
+        for fast cooling:
+        .. math::
+
+            amp = r
+
+            delta = sign(amp-0.5)*T*((1+\\frac{1.0}{T})^{\\abs{2*amp-1}-1.0)
+
+        where :math: `r \\sim \\mathcal{U}(0,1)`
+
+        for cauchy cooling:
+        .. math::
+
+            amp = r
+
+            delta = alpha * T * tan(amp)
+
+        where :math: `r \\sim \\mathcal{U}(-\\pi,\\pi)`
+    """
+    if cont:
```

Comment:
this is not needed... since the else is a "todo"

Reply:
not addressed

Reply:
resolved